wallet/dapps-communication.md

```diff
@@ -8,4 +8,67 @@ Initial meta issue: <https://github.com/massalabs/massa-standards/issues/13>
 
 ## Specification
 
+Here is pseudo code for the two points of view:
+
+- browser extension developer
+- dapp builder
+
+**As a browser extension developer, I would write...**
+
+browser extension code:
+
+```text
+// the extension register itself
+window.massaWalletProvider.addEventListener('loaded', () => {
+  window.massaWalletProvider.emit('register', {
+    ProviderName: "bearby",
+    htmlObject: "window.bearyWalletProvider"
+  })
+})
+```
+
```

Comment:
Here is missing the logic for the bearby wallet to:
- return wallets (listWallets function)
- for a given wallet, return Address, balance and sign command

Reply:
The wallet functionality I presume needs to be concealed inside the injected js into the dom (I have suggested this here: https://github.com/massalabs/massa-standards/pull/26/files#diff-dad455c0814fa2229eaae244a61b26f2f312670c0d14b6da452cd46744739956R73). It contains all the funcs we need the interface to implement.

Reply:
No, my question is more: how in practice the browser extension developer will returnthe address or the balance of a wallet ? Sign a transaction ? Can you propose some pseudo code explaining that ?

Reply:
If I were to request the balance of a wallet inside the code, I would do:

```typescript
// if the window object has the massa object and a method to get all wallet providers
if (window.massa && window.massa.listWalletProviders) {
    // get all wallet providers
    const walletProviders = window.massa.listWalletProviders();
    // choose which provider to use
    const provider = walletProviders[0]; // get some provider based on the metadata e.g.
    // get wallet from that provider
    const wallet = provider.getWallet();
    // use the wallet methods ...
    wallet.getAddress(....)
}
```
As already explained, the Provider itself will communicate with the extension over message dispatching using a controller and `window.postMessage` method. Once the message is posted, the wallet plugin (e.g. a chrome plugin) will have a listener attached listening for messages of a given type and payload and if the message is designated for the plugin (based on the metadata), the plugin will respond to it and send the result back. I will present a schema explaining all of that.

Reply:
@BatiGencho This code is from the Dapps dev point of view.

Can you put yourself in the shoes of the extension developer and write the very same code to handle "getWallets" and "getAddress" ?

can you write the pseudo code doing: 

>have a listener attached listening for messages of a given type and payload and if the message is designated for the plugin (based on the metadata), the plugin will respond to it and send the result back.

and the one doing:

> Provider itself will communicate with the extension over message dispatching using a controller and window.postMessage method

by considering my previous comment (about events on `window` object) ?

Reply:
when registered, the bearby wallet will emit an event with the eventTarget name, so that the massa library can create a wallet object. So in what I am proposing, the wallet object is not defined in bearby side but in massa library side. The wallet object is a class with a sign method. (see bellow)